src/fhirtypes/StructureDefinition.ts

```diff
@@ -158,26 +157,72 @@ export class StructureDefinition {
 
   /**
    * Adds an ElementDefinition to the StructureDefinition's elements, inserting it into the proper location based
-   * on its ID.  This should be used rather than pushing directly to the elements array.
+   * on its ID and sliceName.  This should be used rather than pushing directly to the elements array.
    * @param {ElementDefinition} element - the ElementDefinition to add
    */
   addElement(element: ElementDefinition) {
-    let i = 0;
-    let lastMatchId = '';
-    for (; i < this.elements.length; i++) {
-      const currentId = this.elements[i].id;
-      if (new RegExp(`^${escapeRegExp(currentId)}[.:/]`).test(element.id)) {
-        lastMatchId = currentId;
-      } else if (
-        !new RegExp(`^${escapeRegExp(lastMatchId)}[./:]`).test(currentId) ||
-        // If element is not a slice at this level, and the currentId is a slice, break to add children before slices
-        (new RegExp(`^${escapeRegExp(lastMatchId)}[.]`).test(element.id) &&
-          new RegExp(`^${escapeRegExp(lastMatchId)}[:/]`).test(currentId))
-      ) {
-        break;
+    element.structDef = this;
+    this.addElementToTree(element);
+    if (element.treeParent?.treeChildren != null) {
+      if (element.sliceName != null) {
+        // a newly added element with a slicename should have at least one sibling: its sliced element
+        // all slices come after all child elements.
+        // so, insert a slice before the sliced element's next non-slice sibling that comes after it
+        // this is unfortunately most easily done with string tests since we need to keep slices and reslices organized
+        const slicedElement = element.slicedElement();
+        let i = slicedElement ? this.elements.indexOf(slicedElement) : 0;
+        let lastMatchId = this.elements[i].id;
+        for (; i < this.elements.length; i++) {
+          const currentId = this.elements[i].id;
+          if (new RegExp(`^${this.escapePath(currentId)}[.:/]`).test(element.id)) {
+            lastMatchId = currentId;
+          } else {
+            const lastMatchRegex = this.escapePath(lastMatchId);
+            if (
+              !new RegExp(`^${lastMatchRegex}[./:]`).test(currentId) ||
+              // If element is not a slice at this level, and the currentId is a slice, break to add children before slices
+              (new RegExp(`^${lastMatchRegex}[.]`).test(element.id) &&
```

Comment:
This is very nitpicky, but since you're only matching on one character, I think this would suffice (and might be more efficient depending on RegEx implementation):
```suggestion
              (new RegExp(`^${lastMatchRegex}\\.`).test(element.id) &&
```